### PR TITLE
Add average discounted episodic return metric

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
This metric computes the average discounted episodic return. 
It can be viewed as an approximate visualization of ```E_{s} V(s)```.

### CODE REVIEW

File: alf/metrics/metrics.py [hunk 1]
```diff
@@ -266,6 +279,68 @@ def _extract_metric_values(self, time_step):
...
+
+        super(AverageDiscountedEpisodicReturnMetric, self).__init__(
+            name=name,
+            dtype=dtype,
+            prefix=prefix,
+            batch_size=batch_size,
+            buffer_size=buffer_size,
+            example_metric_value=example_metric_value)
+
+    def _extract_metric_values(self, time_step):
```
Comment (emailweixu) [inline alf/metrics/metrics.py:334]: So this metric will only calculate the discounted return at the beginning of an episode. The values for RL algorithms usually means the average discounted return over all states. So that are not directly comparable.

File: alf/metrics/metrics.py [hunk 2]
```diff
@@ -266,6 +279,68 @@ def _extract_metric_values(self, time_step):
...
+        if time_step.reward.ndim == ndim:
+            return time_step.reward * self._accumulated_discount
+        else:
+            reward = time_step.reward.reshape(*time_step.step_type.shape, -1)
+            return [
+                reward[..., i] * self._accumulated_discount
+                for i in range(reward.shape[-1])
+            ]
+
+    def _after_update(self, time_step):
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: I don't see the need to introduce another hook. Why don't you do this inside _extract_metric_values()?

Comment (Haichao-Zhang) [reply alf/metrics/metrics.py]: Originally to make different functionalities more isolated. I can merge them.

File: alf/metrics/metrics.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py:334]: That is true. The current metric is episodic and not directly comparable to the ``values`` curve.
I'm also trying to implementing a version by averaging over the states encountered during rollout.

Comment (le-horizon) [reply alf/metrics/metrics.py:334]: You'll probably need to record the value computed during loss computation, store it into some member variable of the algorithm, and then plot it (alf.summary.scalar/histogram) in summarize_rollout().

File: alf/algorithms/rl_algorithm.py [hunk 3]
```diff
@@ -147,6 +147,10 @@ def __init__(self,
                     batch_size=env.batch_size,
                     buffer_size=metric_buf_size,
                     reward_shape=env.reward_spec().shape),
+                alf.metrics.AverageDiscountedEpisodicReturnMetric(
+                    batch_size=env.batch_size,
+                    buffer_size=metric_buf_size,
+                    reward_shape=env.reward_spec().shape),
```
Comment (le-horizon) [inline alf/algorithms/rl_algorithm.py]: maybe put at the end, after EnvInfo metric?

File: alf/metrics/metrics.py [hunk 4]
```diff
@@ -266,6 +279,68 @@ def _extract_metric_values(self, time_step):
...
 
+class AverageDiscountedEpisodicReturnMetric(AverageEpisodicSumMetric):
+    """Metric for computing the average discounted episodic return."""
+
+    def __init__(self,
+                 name='AverageDiscountedEpisodicReturnMetric',
+                 prefix='Metrics',
+                 reward_shape=(),
+                 dtype=torch.float32,
+                 discount=0.99,
```
Comment (le-horizon) [inline alf/metrics/metrics.py:312]: it may be better to get this discount from algorithm's loss, if possible, so they are consistent?

File: alf/metrics/metrics_test.py [hunk 5]
```diff
@@ -68,12 +69,17 @@ def _create_trajectories(self):
...
          ('testAverageEpisodeLengthGraph', AverageEpisodeLengthMetric, 6, 2.0),
          ('testAverageEnvInfoMetric', AverageEnvInfoMetric, 6,
           dict(x=torch.as_tensor(5.), y=torch.as_tensor(1.5)))])
     def testMetric(self, metric_class, num_trajectories, expected_result):
         trajectories = self._create_trajectories()
-        if metric_class in [AverageReturnMetric, AverageEpisodeLengthMetric]:
+        if metric_class in [
+                AverageReturnMetric, AverageDiscountedEpisodicReturnMetric,
+                AverageEpisodeLengthMetric
+        ]:
```
Comment (le-horizon) [inline alf/metrics/metrics_test.py]: I'd be happier if a test for multi dim reward could be added, since this change handles multi dim reward as well.

File: alf/trainers/policy_trainer.py [hunk 6]
```diff
@@ -358,6 +358,9 @@ def __init__(self, config: TrainerConfig):
                 alf.metrics.AverageReturnMetric(
                     buffer_size=self._num_eval_episodes,
                     reward_shape=self._eval_env.reward_spec().shape),
+                alf.metrics.AverageDiscountedEpisodicReturnMetric(
+                    buffer_size=self._num_eval_episodes,
+                    reward_shape=self._eval_env.reward_spec().shape),
```
Comment (le-horizon) [inline alf/trainers/policy_trainer.py]: same here, move to last?

**Review by le-horizon — COMMENTED** [review]
Hi Haichao,

I don't fully understand the need for this change.  If we are trying to figure out whether value overestimates, then we should maybe look at rollout value or training losses for value?
Other than that, this change looks good in general.  A few suggestions inlined.

Thanks,
Le

File: alf/metrics/metrics.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py]: Removed this function.

File: alf/metrics/metrics.py (hunk 1 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py:334]: > So this metric will only calculate the discounted return at the beginning of an episode. The values for RL algorithms usually means the average discounted return over all states. So that are not directly comparable.

Now changed the metric to be the average of all the states in rollout.

Comment (Haichao-Zhang) [reply alf/metrics/metrics.py:334]: > You'll probably need to record the value computed during loss computation, store it into some member variable of the algorithm, and then plot it (alf.summary.scalar/histogram) in summarize_rollout().

Calculating the value during rollout is probably the best to be comparable to the discounted return. This is not included currently as it will have some additional computational cost. 
Currently, we may use the the value from the loss function as a coarse estimation.

File: alf/algorithms/rl_algorithm.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/algorithms/rl_algorithm.py]: Done

File: alf/metrics/metrics.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py:312]: It is not straightforward to retrieve the discount value. Now make it configurable so we can set it as we wish.

File: alf/metrics/metrics_test.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics_test.py]: Added

File: alf/trainers/policy_trainer.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/trainers/policy_trainer.py]: Done

**Comment by Haichao-Zhang** [comment]
> Hi Haichao,
> 
> I don't fully understand the need for this change. If we are trying to figure out whether value overestimates, then we should maybe look at rollout value or training losses for value?
> Other than that, this change looks good in general. A few suggestions inlined.
> 
> Thanks,
> Le

This PR is summarize the discounted return (based on rollouts), which is what the value function tries to predict.

You are right, calculating the value during rollout is probably the best to be comparable to the discounted return in order to determine the degree to which we have overestimation. 
This is not included currently as it will have some additional computational cost.
Currently, we may use the the value from the loss function as a coarse estimation.

File: alf/metrics/metrics.py [hunk 7]
```diff
@@ -224,18 +214,32 @@ def _update_accumulator_(acc, val):
...
+                lambda buf, acc: buf.append(
+                    self._extract_and_process_acc_value(
+                        acc, last_episode_indices)), self._buffer,
+                self._accumulator)
 
         return time_step
 
+    def _extract_and_process_acc_value(self, acc, last_episode_indices):
+        """Extract the final accumulated value and perform some optional
+        customizable processings.
```
Comment (le-horizon) [inline alf/metrics/metrics.py]: processings => processing

File: alf/metrics/metrics.py (hunk 1 above)
Comment (le-horizon) [reply alf/metrics/metrics.py:334]: Ah, you are right, my suggestion won't work for rollout loss.

**Review by le-horizon — DISMISSED** [review]
Thanks for the improvements and quick turnaround, Haichao.  LGTM!

Le

File: alf/metrics/metrics.py [hunk 8]
```diff
@@ -266,6 +283,107 @@ def _extract_metric_values(self, time_step):
...
+    It is calculated according to the following formula:
+
+    .. math::
+        \begin{array}{ll}
+            R &= r_1 + (1 + gamma) r_2 + + (1 + gamma + gamma^2) r_3 + \cdots \\
+            &= \frac{1}{L}\sum_{l=1}^L \sum_{k=0}^{l-1} \gamma^k r_l,
+        \end{array}
+    where :math:`\gamma` is the reward discount.
+
+    Note that if the last step is not due time limit, the discounted return
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: due => due to

File: alf/metrics/metrics.py [hunk 9]
```diff
@@ -266,6 +283,107 @@ def _extract_metric_values(self, time_step):
...
+    calculated form the formula above is unbiased. If the last step is due
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: due to

File: alf/metrics/metrics.py [hunk 10]
```diff
@@ -266,6 +283,107 @@ def _extract_metric_values(self, time_step):
...
+        second step. Otherwise, is it updated by multiplying with
+        ``self._discount``.
+        Args:
+            time_step (alf.data_structures.TimeStep): batched tensor
+        """
+        is_first = time_step.is_first()
+
+        # update discount for the next time step
+        self._accumulated_discount = self._accumulated_discount + \
+                            self._discount**self._current_step
```
Comment (emailweixu) [inline alf/metrics/metrics.py]: A slightly faster version:
```python
self._accumulated_discount = self._discount * self._accumulated_discount + 1
```

File: alf/metrics/metrics.py [hunk 11]
```diff
@@ -266,6 +283,107 @@ def _extract_metric_values(self, time_step):
...
+        self._accumulated_discount = torch.where(
```
Comment (emailweixu) [inline alf/metrics/metrics.py:372]: For FIRST, this is 0, which means that reward for FIRST step is ignored (which is correct, BTW). Because of this, at line 383, should we divide acc using (self._current_step[last_episode_indices] - 1) instead?

Comment (Haichao-Zhang) [reply alf/metrics/metrics.py:372]: The ```_current_step``` is initialized to be zero for FIRST step and is increased by one each time step after that.
Therefore, at the episode end, ```self._current_step[last_episode_indices]```  equals to episode_length-1.
So it actually performs the correct calculation as you suggested.

I added more comments on this.

File: alf/metrics/metrics.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py]: fixed

File: alf/metrics/metrics.py (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py]: fixed

File: alf/metrics/metrics.py (hunk 9 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py]: fixed

File: alf/metrics/metrics.py (hunk 10 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py]: Changed

File: alf/metrics/metrics.py [hunk 12]
```diff
@@ -266,6 +283,114 @@ def _extract_metric_values(self, time_step):
...
 
 
+@gin.configurable
+class AverageDiscountedReturnMetric(AverageEpisodicSumMetric):
+    """Metric for computing the average discounted episodic return.
+    It is calculated according to the following formula:
+
+    .. math::
+        \begin{array}{ll}
+            R &= r_1 + (1 + gamma) r_2 + (1 + gamma + gamma^2) r_3 + \cdots \\
```
Comment (hnyu) [inline alf/metrics/metrics.py]: gamma needs backslash; just seen from this first line, it's not the discounted return; an average is needed

File: alf/metrics/metrics.py [hunk 13]
```diff
@@ -266,6 +283,114 @@ def _extract_metric_values(self, time_step):
...
+            &= \frac{1}{L}\sum_{l=1}^L \sum_{k=0}^{l-1} \gamma^k r_l,
```
Comment (hnyu) [inline alf/metrics/metrics.py:294]: Do we want to average over the ending steps or the beginning steps? If this quantity is to be compared to the value estimation, then we want to average over the beginning steps. In the episodic setting:

\frac{1}{L}\sum_{l=1}^L \sum_{k=l}^{L} \gamma^{k-l} r_k,

File: alf/metrics/metrics.py [hunk 14]
```diff
@@ -266,6 +283,114 @@ def _extract_metric_values(self, time_step):
...
+        \end{array}
+    where :math:`\gamma` is the reward discount, and :math:`r_1` denotes the
+    reward due to the first action, which is received at the second time step.
+    :math:`L` equals to the episode length - 1.
+
+
+    Note that if the last step is not due to time limit, the discounted return
+    calculated form the formula above is unbiased. If the last step is due to
+    time limit, it is a biased estimate and its expectation is lower than the
```
Comment (hnyu) [inline alf/metrics/metrics.py:303]: If the last time step is TimeLimit, then I think a better metric would be still using the last step's value estimation for the remaining sum.

File: alf/metrics/metrics.py (hunk 12 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py]: Fixed.

File: alf/metrics/metrics.py (hunk 14 above)
Comment (Haichao-Zhang) [reply alf/metrics/metrics.py:303]: Agree. In this case, the metric is not accurate.

Comment (hnyu) [reply alf/metrics/metrics.py:303]: For the last several steps, this bias is very severe. If you're going to compare the average metric with the predicted discounted value, then probably the comparison is not informative. Imagine that the predicted value is around 100 at every step in the non-episodic setting, but the steps towards the time limit only have accumulated rewards of about 1.

**Review by emailweixu — APPROVED** [review]
